### PR TITLE
Tags & Topics

### PR DESCRIPTION
**Pull request by UladBohdan**
### Topics & Tags
* `Topic` model is introduced; all topics (their slugs) are now returned as a part of a `GET /api/storage/mainPage` request
* Note `topic` tab in our i18n google spreadsheet
* `Tag` model is introduced together with validation rules (Joi) for its content
* `GET /api/tags/by-topic/locations` - returns all *location tags*
* `GET /api/tags/articles/bykau` - returns all *articles* which are associated with Vasil Bykau
* `tags` are an array field in `Article` model
* script to fetch **Tags** from Google Spreadsheet and to push them to database
https://docs.google.com/spreadsheets/d/1GxVDSZOWyGvtXa_dwmaRcjzWrXq_kjgXCmZEJVrgb10
Basic usage: `npm run fetch-tags -- secret-dev.json`

### Minor changes:
* `color` (hex) and `textColorTheme` (enum [`dark`, `light`]) are added to Article Model.

### Also:
* `@babajka.io` is now a preferred kind of emails in testing; `@wir.by` emails are now only used for generated authors and tests for them
* flaky auth tests seem to be fixed
* a minor change for Storage Endpoints: they are `GET /api/storage/main-page` and `POST /api/storage/main-page` now

### CODE REVIEW

**Comment by UladBohdan** [comment]
[to add a general-purpose "Tag" model to backend and develop API](https://trello.com/c/dH6xsnSW/302-to-add-a-general-purpose-tag-model-to-backend-and-develop-api)

**Comment by UladBohdan** [comment]
[to unify all @babajka.io / @wir.by occurrences in tests](https://trello.com/c/0hFjVnsf/231-to-unify-all-babajkaio-wirby-occurrences-in-tests)

File: src/db/scripts/fetchTags.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  new Promise(resolve => {
+    const result = [];
+    sheet.getRows({}, (error, rows) => {
+      rows.shift(); // This removes the "comment" row.
+      rows.forEach(row => {
+        const data = {};
+        scopesUnlocalized.forEach(scope => {
+          const value = row[scope];
+          if (value) {
+            set(data, [scope], row[scope]);
```
Comment (Drapegnik) [inline src/db/scripts/fetchTags.js]: why `[scope]`?

Comment (Drapegnik) [inline src/db/scripts/fetchTags.js]:
```suggestion
            set(data, [scope], value);
```

File: src/db/scripts/fetchTags.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+          }
+        });
+        locales.forEach(locale => {
+          scopesLocalized.forEach(scope => {
+            const value = row[`${scope}.${locale}`];
```
Comment (Drapegnik) [inline src/db/scripts/fetchTags.js]: you can use lodash `get`

File: src/db/scripts/fetchTags.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    throw new Error('it seems like there are slug duplicates in the Spreadsheet');
+  }
+
+  return allTags;
+};
+
+const updateTags = async (sheetsData, db) => {
+  const promises = [];
+
+  Object.entries(sheetsData).forEach(([topic, tags]) => {
```
Comment (Drapegnik) [inline src/db/scripts/fetchTags.js]: please use `map` instead `forEach` & `push`

File: src/db/scripts/fetchTags.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    )
+  );
+
+  return data;
+};
+
+const validateSheetsData = sheetsData => {
+  const allTags = [];
+
+  Object.entries(sheetsData).forEach(([topic, tags]) => {
```
Comment (Drapegnik) [inline src/db/scripts/fetchTags.js]: please use `map` instead `forEach` & `push`

File: src/api/article/article.model.js [hunk 5]
```diff
@@ -86,6 +86,27 @@ const ArticleSchema = new Schema(
       // Can only be present when Article type is video.
       type: VideoReferenceSchema,
     },
+    color: {
+      // Articles have colors to be rendered on the main page.
+      type: String,
+      validate: colorValidator,
+      required: true,
+      default: '000000',
```
[the remark this answers is not in the document]
Comment (UladBohdan) [reply src/api/article/article.model.js]: I don't know. I can't see any reasons to remove `required: true` option either.

File: src/api/tag/tests.js [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+const request = supertest.agent(app.listen());
+
+describe('Tag model', () => {
+  let defaultMetadata;
+  let topics;
+
+  // eslint-disable-next-line func-names
+  before(async function() {
+    // I should probably avoid these lines.
+    this.timeout(5000);
```
[the remark this answers is not in the document]
Comment (UladBohdan) [reply src/api/tag/tests.js]: I forgot about these lines😬

But using timeouts in tests is not a good thing to do. But that makes them non-flaky here so I would leave it.

File: src/auth/tests.js [hunk 7]
```diff
@@ -11,6 +11,9 @@ app.get('/protected', requireAuth, (req, res) => res.sendStatus(200));
 
 describe('Auth API', () => {
   before(async () => {
+    // It is not clear why dropData() here helps with flaky auth tests. But it helps.
```
[the remark this answers is not in the document]
Comment (UladBohdan) [reply src/auth/tests.js]: That's a workaround:(
There's a card though: https://trello.com/c/dN5Jx7yt/339-to-replace-all-afterdropdata-occurrences-with-before

File: src/db/scripts/fetchTags.js (hunk 1 above)
Comment (UladBohdan) [reply src/db/scripts/fetchTags.js]: made it `scope`. Doesn't really matter, both are fine.
https://lodash.com/docs/4.17.11#set

File: src/db/scripts/fetchTags.js (hunk 2 above)
Comment (UladBohdan) [reply src/db/scripts/fetchTags.js]: it's not the same. I can't use it as `get(row, [scope, locale])` if that's what you mean

File: src/db/scripts/fetchTags.js (hunk 3 above)
Comment (UladBohdan) [reply src/db/scripts/fetchTags.js]: The problem was the arrays were nested.
Rewrote it with `map`-s and `flatten`.

**Pull request merged**